### ISSUE TITLE
stack/ustack to accept limit argument

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#11]
stack and ustack are currently implemented as builtins, and return the full kernel and user stack traces. This ticket is to change them to accept an argument: the top number of frames to print. So instead of "stack" for the full stack, we can use "stack(3)" to print the top three frames only, and "stack()" to print all frames.

This should work:

```
# ./src/bpftrace -e 'kprobe:decay_load { @[stack(3)] = count(); }'
BDG: value_size 8
Attaching 1 probe...
[...]
]: 172
@[
decay_load+1
update_load_avg+1609
enqueue_entity+104
]: 172
@[
decay_load+1
update_blocked_averages+1334
run_rebalance_domains+114
]: 173
```

So that's showing the traced function (decay_load) and two more frames only, for the maximum of 3 frames.

This will mean changing them from a builtin to a call, and pulling in the argument in ast/codegen_llvm.cpp. There's more than one way to pull in an argument. Eg:

    call.vargs->front()->accept(*this);   // it becomes expr_

or, from lhist:

    Integer &value_arg = static_cast<Integer&>(*call.vargs->at(0));
    Integer &min_arg = static_cast<Integer&>(*call.vargs->at(1));
    Integer &max_arg = static_cast<Integer&>(*call.vargs->at(2));
    Integer &step_arg = static_cast<Integer&>(*call.vargs->at(3));

Now the tricky part: how do we limit frames? Ideally, this is done in-kernel. However, I don't know how we do that. I've discussed it in:

https://github.com/iovisor/bcc/issues/1103

So this might involve editing CreateGetStackId from ast/irbuildirbpf.cpp, or, it might involve setting and restoring the sysctl kernel.perf_event_max_stack. Or something else: maybe this requires a kernel change to support it (the reserved bits).

stack()/ustack() with no arguments should print the full stack.

Later on (and as a separate ticket) we can add a 2nd argument for skip frames, which bpf_get_stackid() supports.

**Comment by brendangregg** [comment]
I raised this on the call today and @4ast wasn't sure there was a way but maybe there was (other than setting it system wide using the sysctl kernel.perf_event_max_stack). We're going to have to study the kernel source know for sure if it's doable now or not.

**Comment by 4ast** [comment]
@brendangregg when stackmap is created value_size gives the limit (in bytes) on number of elements in the stack trace.

**Comment by brendangregg** [comment]
Great, thanks @4ast !

**Comment by brendangregg** [comment]
This is an api change, from ustack to ustack([int limit]). Adding 1.0 tag.

**Comment by brendangregg** [comment]
I've dug into this, and it's more involved. Consider this program:

```
k:vfs_read
{
    @vfs = stack(4);
}

k:vfs_write
{
    @vfs = stack(2);
}

k:do_nanosleep
{
    @n = stack(8);
}

k:tcp_sendmsg
{
    @t = stack(4);
}
```

The k:vfs_write should fail in semantic analyser because `@vfs` was declared as a 4 frame stack, and it needs to stay that way in subsequent uses. This shouldn't be too hard to enforce.

But the k:do_nanosleep stack presents a problem: `@vfs` wants to hash on 4 frames, whereas `@n` wants to hash on 8. There's only one stack map in bpftrace:

```
    if (needs_stackid_map_)
      bpftrace_.stackid_map_ = std::make_unique<bpftrace::FakeMap>(BPF_MAP_TYPE_STACK_TRACE);
```

But now there needs to be multiple stack maps, since they are hashing on different numbers of frames. Perhaps we need a stack map for each map, or a stack map for each number of frames used. If it's the latter, then `@vfs` and `@t` can share the same stack map.

I might fix this after the stack to kstack rename (#344) since it'll touch similar code (in this case, I'd be moving kstack from a builtin to a call).

**Comment by mmarchini** [comment]
I'll try to implement this. Should we keep both `ustack` and `ustack([int limit])`?

**Comment by brendangregg** [comment]
Thanks! That's an interesting idea. We currently have builtins and calls, so if we allowed both then ustack would look like both a builtin and a call. I'm not sure if that would be too confusing for end users. But on the other hand, why would we be forced to type "()" needlessly if we weren't providing a number. I don't have a strong opinion.

I did try out Alexei's suggestion and it worked, so I know this is doable. From memory I was adjusting MAX_STACK_SIZE in map.cpp. To fix this issue, we'll want a separate stack map for each use of ustack or kstack, unless they have the same number of frames (although to start with, just having separate maps for each would work).

The default MAX_STACK_SIZE should be 128 (not 32), and we should support tuning it up to 1024 at least (I've seen stacks that exceed that). I just noticed "max_entries" in map.cpp 128 for stack maps: that should be 2048 at least.
